### ISSUE TITLE
Feature: Automatic install of extension dependencies

### ISSUE THREAD

**Issue by brpaz** [Ulauncher/Ulauncher#273]
Hello. 
I think the most common issue people have when installing extensions is because of missing dependencies.

It would be nice if Ulauncher could install dependencies automatically. 
Probably the simplest and more native way, since Ulauncher is built with Python  would be to use requirements.txt file. Ulauncher could detect if the requirements.txt file is present In the extension soure code and if yes it would run ```pip install -r requirements.txt```  automatically before starting the extension.

**Comment by gornostal** [comment]
Hi @brpaz 

Thanks for the suggestion. I understand the problem with the missing dependencies and approach you've described should work.

**Comment by friday** [comment]
For a CLI dependency the below may work. Not really optimal for more complex dependencies, or if it's critical that they get the most recent version, though you could probably adapt it to work for some of these cases too.

```python
import os
from distutils.spawn import find_executable

cmd = 'binary_name'
extDir = os.path.dirname(os.path.realpath(__file__))

# test if client is installed on system and is in PATH or change to local version
if not find_executable(cmd):
    cmd = '{}/{}'.format(extDir, cmd)

# test/get local version
if not find_executable(cmd):
    # Simple example assuming the url is directly to a binary. More likely you'll get an archive or a .deb which needs to be extracted
    urllib.urlretrieve('https://...', cmd)
    os.chmod(cmd, 0o755)
```

**Comment by brpaz** [comment]
Think this is an important issue.Most of the issues opened in my extensions repos are because of missing dependencies.

I try to keep the dependencies required listed in the README but it doesn't seem to help much. ;(

We need a more automatic way.

**Comment by friday** [comment]
Yeah, the average user doesn't bother reading instructions it seem. Just posts straight to Disqus that it doesn't work, even if the dependency is mentioned in the title. :P

Would it be feasible to include it in the source? Your Lorem Ipsum extension for example has a tiny dependency, which is not likely that people have. For GPL dependencies you may have to use GPL too this way though. Not sure.

Also, when failing to meet a dependency, you could perhaps show a notification so users become aware of the issue: https://github.com/Ulauncher/ulauncher-kill/blob/master/main.py#L73

**Comment by friday** [comment]
For `pip` this should work:

```python
import subprocess
import sys

def ensure_import(package):
    try:
        __import__(package)
    except ImportError:
        subprocess.call([sys.executable, "-m", "pip", "install", "--user", package])
    return __import__(package)
```

Example:
```
lorem = ensure_import("lorem")
print(lorem.sentence())
```

It will use the correct pip binary for the system.

`ensure_import()` should not be run on demand within the `KeywordQueryEventListener` since it may take a second or so to install the module.

Binaries installed this way will end up in `$HOME/.local/bin/`. This is not likely in path, but you don't really need it to be. You can just run them with that prefix.

**Comment by Riezebos** [comment]
Even if I follow the prereqs of an extension I often still get errors because some packages are assumed to be installed (probably because they are common). I don't use Python 2 at all normally, so I haven't installed any packages there. Usually if I want to use an extension I first have to open main.py and manually run the import statements to see which packages pop up as missing (e.g. requests). To the outsider the Ulauncher extension ecosystem can look like it's an unmaintained mess because half the extensions throw a weird API error (on standard Ubuntu 18.04 at least).

After debugging I personally like the extensions ecosystem a lot!

**Comment by friday** [comment]
> Even if I follow the prereqs of an extension I often still get errors because some packages are assumed to be installed (probably because they are common)

That's their mistake imo. Installing packages globally means you have need sudo access to install extension though, and also indirectly introducing conflicts with packages in the distro.

> To the outsider the Ulauncher extension ecosystem can look like it's an unmaintained mess because half the extensions throw a weird API error (on standard Ubuntu 18.04 at least).

I think the problem here is made worse by failing silently or just logging the error. That is, not doing this: https://github.com/friday/ulauncher-clipboard/pull/8.

**Comment by Riezebos** [comment]
I agree on both points, but even if it's their fault it reflects badly on Ulauncher if it occurs often. Another solution for this would be to test every extension in a controlled environment with just the minimal requirements + stated prerequisites before the extension is added to the website.

**Comment by friday** [comment]
I think a simple solution that works for most extensions is better than spending a lot of time on this and adding much code that needs to be maintained.

Suggestion:
* Add a key/value list or multiple arrays for dependencies in the json manifest. It needs to separate python dependencies from binaries.
* Install the python dependencies locally when installing the plugin, unless the manifest specifies that it shouldn't be installed locally
* Show a warning for any extension that have unmet dependencies, rather than assuming extension devs will handle it (evidently they don't, and as you say, this reflects back on ulauncher).

**Comment by brpaz** [comment]
> Even if I follow the prereqs of an extension I often still get errors because some packages are assumed to be installed (probably because they are common)

Yes, I am guilty of this. I dont have many experience with Python and some modules are used everywhere so i think they are from the standard lib, when in fact they are not.

> Add a key/value list or multiple arrays for dependencies in the json manifest. It needs to separate python dependencies from binaries, and perhaps python modules from python binaries bundled as modules.

Why not using requirements.txt, which is the standard way to define dependencies in Python, instead of adding to json manifest? it wont work for binaries tough. so yeah, I guess your solution of using the manifest would be more complete.

**Comment by friday** [comment]
> Why not using requirements.txt, which is the standard way to define dependencies in Python, instead of adding to json manifest? it wont work for binaries tough. so yeah, I guess your solution of using the manifest would be more complete.

If that's better for devs and/or easier to implement/maintain I think it's reason enough. My main concern is actually showing a warning if a dependency is unmet so they can address the problem themselves. That could be an entry in the manifest while also supporting requirements.txt

**Comment by endorama** [comment]
Hello, I'm currently working on an extension that requires a cli component to be present in order to work, so this is very relevant to me!

I implemented an automatic downloader and I'm adding a notification to alert the user of the missing tool, but I find this a bit of bad UX. As potentially all extensions require this, feels like a single user can potentially be targetted by lots of alerts/notifications when starting ulauncher (depending on how many extensions will start implementing these behaviours).

Would be awesome to have a section in the Ulauncher preference panel where such notification are shown with some additional information, so that extension developers have a standardize way to interact with the user on this. O

**Comment by brpaz** [comment]
Just a compilation of issues opened in my Extensions repositories in the last weeks, all related to missing dependencies.

* [KeyError · Issue #1 · brpaz/ulauncher-cloudflare](https://github.com/brpaz/ulauncher-cloudflare/issues/1)
* [Unexpected API error. · Issue #1 · brpaz/ulauncher-devhints](https://github.com/brpaz/ulauncher-devhints/issues/1)
* [Dependency confusion / Unexpected API error · Issue #3 · brpaz/ulauncher-gitlab](https://github.com/brpaz/ulauncher-gitlab/issues/3)
* [no options available... · Issue #3 · brpaz/ulauncher-pwgen](https://github.com/brpaz/ulauncher-pwgen/issues/3)
* [crashed instantly · Issue #5 · brpaz/ulauncher-pwgen](https://github.com/brpaz/ulauncher-pwgen/issues/5)

This is getting really hard to control.

And while a Python dev can probably figure it out looking the extension logs, It is much harder for end users of Ulauncher that just go to the Extensions Website and install from there.

**Comment by ptandler** [comment]
While the installation of requirements is not yet automated, it would be helpful if the requirements are at least *listed* on the extension page, see #581
If I *know* that there are additional requirements, I can install them manually ...

**Comment by ptandler** [comment]
Well: And an error message complaining about missing requirements would also help of course. Ideally with hints of what is missing and how to install it. I don't remember exactly, but I think I just got a message saying something like "extension terminated" without anything else.

**Comment by tchar** [comment]
Hello,

In case anybody is interested [here](https://github.com/tchar/Ulauncher/tree/requirements) is an implementation of automatic installation of requirements using `venv`s installed inside the extension directory.

**Comment by BharatKalluri** [comment]
There is a project called [pex](https://github.com/pantsbuild/pex), stands for python executable. With that, we can create binaries right from a simple setup.py project which will include all the dependencies and can function independently. 
Going on the same idea, since the code is open source, maybe Ulauncher could run a fortnightly job in a server and git pull all the extensions, run and generate the pex files and populate it for the extensions website. Installing will also be a breeze since all of them are binaries.
Let me know if I'm missing anything here, since I'm not fully aware of the problems Ulauncher ran into during extensions management :)

**Comment by troycurtisjr** [comment]
So I think it is pretty clear that some iteration of using virtual environments is the way to approach this issue. While ulauncher could conceivably automatically install to the user location, using `pip install --user ...`, that is probably really bad form. A few challenges I had initially thought of was that the extension really really needs to be running the exact same version of Ulauncher, and the Ulauncher server. My experience with virtual environments in the past have always started from scratch and it seems like it would be a bit unmanageable, not to mention wasteful, to install the full ulauncher dependency tree in each extensions virtual environment. 

Enter `--system-site-packages`, which configures the virtual environment to look first at the available system packages, and if the dependency is satisfied there, it can use it. Otherwise, it will install the needed package.

So with that particular problem seemingly taken care of, I'm going to work up a change/PR which will do the following:
- After downloading the extension, look for a requirements.txt file
- If a requirements.txt file exists:
    - create a virtual env for that extension
    - Use `pip` within the extensions virtual environment to install all dependencies

Then when starting the extension
- If a virtual environment exists for the extension, use the python executable found there, else use the system python

"Check for Updates" will also update pip installed packages if a virtual environment exists

This way the virtual env is only created if necessary, and the same common packages aren't duplicated unnecessarily. Those common packages will also stay on the same version relative to the Ulauncher server.

Extension authors could include all dependencies, including ones they know exist because they are ulauncher dependencies, or only the "extra" dependencies. Whichever works best for their use case. 

Of course this only handles the Python dependencies, which I think is the main thing users need help with. I would expect that most other cases where some external command is needed, the extension is designed *for* that external command, so it makes no sense for a user to want the extension and not the referenced app. Something like someone installing a firefox bookmark extension, but not having firefox installed.

pex is a cool project, but we really need to point to the system Gtk and Ulauncher for best compatibility. `piprun` has a similar issue, in addition to being to "transient" and requiring Internet connectivity for every launch.

**Comment by friday** [comment]
> Of course this only handles the Python dependencies, which I think is the main thing users need help with. I would expect that most other cases where some external command is needed, the extension is designed _for_ that external command, so it makes no sense for a user to want the extension and not the referenced app. Something like someone installing a firefox bookmark extension, but not having firefox installed.

Not to mention we fundamentally can't and shouldn't auto-install those, so we can only inform/warn. In some cases the extension can bundle them, or download them if they're missing as I posted above, but I doubt many extensions will do this.

I had #805 in mind for external commands (it should be easy to add). I guess we could print that message to the log too.

**Comment by iaacornus** [comment]
hi can I be assigned to this issue? thanks

**Comment by iaacornus** [comment]
hi, just confirming before I start working on it. so the issue is some extensions cannot be installed due to dependency problems. Hence, our solution/approach is to create a virtual environment for extensions and to install the dependencies and extensions there? So, I'd like to clarify if:

- create venv for each extensions, install the dependencies there and the extension
- create venv for all extensions and install all of the dependencies needed by the extensions there

other than these, are there anything i should note while implementing this or any constraints? other than performance and resources. Should i be worried about adding another dependencies in ulauncher? or you are fine if i would add another dependency/module. I'd appreciate it if @troycurtisjr can catch me up with the solutions he tried or any tips i could use in approaching this problem. I just finished checking the src code of ulauncher, and trying to build a plan in approaching this problem, so any tips or directions and notes will be greatly appreciated.

**Comment by gornostal** [comment]
Hi @iaacornus 
Thanks for showing your interest in helping with this! :pray: 
Here are my answer to your questions:

> create venv for each extensions, install the dependencies there and the extension

This is a more preferred way than creating a single venv because different extensions may use different versions of the same dependency. But please create it ONLY IF an extension has external dependencies (i.e. requirements.txt file). There's no need to create venv for all extensions.

> Should i be worried about adding another dependencies in ulauncher?

Yes :slightly_smiling_face:  Try to avoid adding dependencies. It's a bit difficult to add them because they should be installed via the system package manager (apt, yum, pacman, etc.), which has veeeeery limited number of python packages and versions.

Regarding the approach, other than what Troy mentioned [here](https://github.com/Ulauncher/Ulauncher/issues/273#issuecomment-1038542547), I'd suggest to do this in two stages:
1) Create a prototype -- don't worry about code quality or nuances that have straightforward solutions. Share the prototype (link to your branch) with us and any findings or edge cases that you encounter (I'm pretty sure there will be some)
2) Finish the feature

**Comment by iaacornus** [comment]
thank you for those clarifications! I'll be sure to take note of those

**Comment by millerdev** [comment]
Maybe consider using [uv](https://docs.astral.sh/uv/) to manage virtualenvs? To make it easier to use uv (which is itself pip-installable), Ulauncher could create a virtualenv for itself in which it can install its own dependencies without mucking up system packages.

Another idea that came to mind while reading this issue is to package the portion of Ulauncher code that is needed to interface with extensions as a stand-alone pip-installable package so each extension that needs a virtualenv can install it to communicate with the main Ulauncher process. There is some versioning complexity to deal with here, but nothing that Ulauncher doesn't already deal with as it evolves.

**Comment by friday** [comment]
The obvious answer to start with is the native python venv lib https://docs.python.org/3/library/venv.html

If that doesn't work for our requirements we have to use something that has Debian/Ubuntu packages (like python3-virtualenv).

**Comment by iaacornus** [comment]
@friday @millerdev 

> Maybe consider using [uv](https://docs.astral.sh/uv/) to manage virtualenvs? To make it easier to use uv (which is itself pip-installable), Ulauncher could create a virtualenv for itself in which it can install its own dependencies without mucking up system packages.
> 
> Another idea that came to mind while reading this issue is to package the portion of Ulauncher code that is needed to interface with extensions as a stand-alone pip-installable package so each extension that needs a virtualenv can install it to communicate with the main Ulauncher process. There is some versioning complexity to deal with here, but nothing that Ulauncher doesn't already deal with as it evolves.

I think I will use the native python venv lib first; if it fails, then perhaps I'll switch into that. My current plan is to write a utility/module that will interact with the environment and manage it. So the outline is:

1. create a venv for each extension, if it has a requirements.txt
2. so we record the requirements.txt that extension and the name of env created for it (perhaps a hashmap?)
3. should there be another extension that uses the same deps/version of dep, install that extension in the same env; otherwise, go back to step 1
4. if it doesnt have requirements.txt, then just create a venv solely for it.

I'll just use venv for here, but that would be a very messy solution, so I'll make a module just for it, and try to not add another dep(s) for ulauncher. If you have any thoughts on this, just drop it here. The branch is iaacornus/Ulauncher/tree/feature-1.

I'll send timely update here as I work through the feature, but expect that it will be a messy code and approach, I'll fix it tho once it works and we agreed on the prototype.

**Comment by friday** [comment]
> 1. create a venv for each extension, if it has a requirements.txt

:+1: 

> 2. so we record the requirements.txt that extension and the name of env created for it (perhaps a hashmap?)

I have no opinion here since I don't know/use venvs. If it's technically required to install/maintain/update the venvs to store this somewhere we should, but otherwise please go for a MVP approach to get it working and should we realize that we need/want these details.

> 3. should there be another extension that uses the same deps/version of dep, install that extension in the same env; otherwise, go back to step 1

I don't know about this :thinking:

If you have two extensions that use the same dep(s), then one extension updates and add more deps, or changes the dependencies/dependency versions, then this seems to makes things more complicated. And also, the locations of the venvs would have to be some random hash then, rather than putting the .venvs in the the extension id/path. I think the point Troy made was rather to check if the system python has the required pip dependencies already and avoid creating a venv then, by using the `--system-site-packages` flag (but again, I know very little about venvs).

> 4. if it doesnt have requirements.txt, then just create a venv solely for it.

If it doesn't have requirements.txt, can't we run directly on the system python?

> If you have any thoughts on this, just drop it here. The branch is iaacornus/Ulauncher/tree/feature-1.
> 
> I'll send timely update here as I work through the feature, but expect that it will be a messy code and approach, I'll fix it tho once it works and we agreed on the prototype.

Please create a draft PR once you have some code (or earlier if you want), and we can discuss the implementation details in there. This issue is very old and has many participants that will get messages every time we post here.

**Comment by iaacornus** [comment]
#1430

**Comment by gornostal** [comment]
I've been researching options. In theory, the following approach should work:

Run this when installing `extension-X`
```
pip install -r extension-X/requirements.txt --target extension-X/deps
```

and then in the extension entry point:

```
import sys
sys.path.insert(0, 'extension-X/deps')

import <some package from requirements.txt>
```

I will do a prototype soon, unless anyone sees any issues with this approach.

**Comment by gornostal** [comment]
The initial implementation is done in v6 branch. Closing this.

**Comment by gerardbosch** [comment]
> The initial implementation is done in v6 branch. Closing this.

Thanks for adding this @gornostal. Is there any plan or roadmap for the v6 to be released anytime soon? Thanks!

**Comment by friday** [comment]
> > The initial implementation is done in v6 branch. Closing this.
> 
> Thanks for adding this @gornostal. Is there any plan or roadmap for the v6 to be released anytime soon? Thanks!

No date, but we're trying to finish off the last tasks. See the pinned issue and V6 milestone
